src/plugins/sync_to_async.ts

```diff
@@ -0,0 +1,380 @@
+    // Check whether this expression is a UI Widget call.
+    if (isUIWidgetExpression(name)) {
+      // Set a flag so that we can later inject the UI Widget
+      // JavaScript/CSS, and a helper to initialise the library.
+      cache.set("uiWidgets", true);
+      return;
+    }
+    // Because we're looking for either v1 api calls, or calls to
+    // other app methods, we only care about ThisExpression's.
+    if (!path.get("object").isThisExpression()) return;
+    // `op` will be the containing app method definition, one of
+    // the direct children of the `container`.
+    const op = path.findParent(path => {
+      return path.isObjectProperty() && cache.has(path.node.key.name);
+    });
+    // Short-circuit if we're not inside an app method, we may be inside
+    // a closure-local method defined in the v1 IIFE.
+    if (!op) return;
+    const opName = op.node.key.name;
+    // Get the cache defined previously in the visitor for
+    // (ObjectMethod|ObjectProperty)
+    const opCache = cache.get(opName);
+    // If we have a methodName defined then we're in the second phase/pass
+    // looking for references to other app methods.  If we're inside the same
+    // method we're looking for, short-circuit.
+    if ((methodName && name !== methodName) || opName === methodName) return;
+    // Get the entire expression we want to replace.  `exp` will be the path
+    // object that wraps the entire expression, i.e. `this.ticket().requester().email()`.
+    // `names` will be an array of names in the expression, i.e. ["ticket", "requester", "email"]
+    const { exp, names } = getExpressionToReplace(path);
+    // If this expression is a call to an app method, and that method is async (already)
+    // and the parent of the exp isn't wrapped in an await, then wrap, i.e.
+    // If we have `{ foo: async function() { ... }, bar: function() { var foo = this.foo(); } }`
+    // and we're inside `bar`, then `exp` will be `this.foo()`.  We'll replace
+    // with `await this.foo()` (and later make `foo` async too).
+    if (
+      (obj = cache.get(name)) &&
+      obj.async &&
+      !exp.parentPath.isAwaitExpression()
+    ) {
+      exp.replaceWith(types.awaitExpression(exp.node));
+      exp.skip();
+      toAsync = true;
+    } else if (apis.has(name)) {
+      // If `name` appears in `apis`, it will be a call to a v1 api.
+      // Next, try work out whether it is a get, set, or invoke (depending
+      // on whether there were any arguments passed to the exp).
+      if (exp.node.arguments.length) {
+        let nexp;
+        // If the method is one of ticketFields, userFierlds or organizationFields
+        // then we need to create a v2 path that uses the colon-delimited style, i.e.
+        // `ticketFields:brand`
+        if (/^(ticket|user|organization)Fields$/.test(name)) {
+          const fieldName = `${name}:${exp.get("arguments.0").node.value}`;
+          nexp = buildWrapZafClientExpression(fieldName);
```

Comment:
How are we handling the ability to set a custom field? e.g. `this.ticket().customField('foo', 'bar')` -> `client.set('ticket.customField:foo', 'bar')`.